### PR TITLE
SpellAbilityAi: move willPayUnlessCost to EffectSpecific Code

### LINKED ISSUE #6421: Volatile Stormdrake bug - opponent doesn't pay energy counters but doesn't sacrifice creature.
Volatile Stormdrake enters, targets 5 MV creature (Marneus Calgar), player gets 4 energy counters, player doesn't pay any, creature isn't sacrificed. No object on board visible that would prevent the sacrifice effect.

Log
image

Boardstate
image

Forge updated latest version.

### PR DESCRIPTION
**Pull request by Hanmac**
This is for moving "willPayUnlessCost" to Effect Specific AI code

with that we will clean up when AI should not Pay for Unless Cost

It is still work in Progress

Closes #6421

### CODE REVIEW

**Comment by Hanmac** [comment]
Probably next one to refactor is LifeLE cost ones.

While not UnlessCost, AI need help with Effects like `Braids, Arisen Nightmare` to not kill themselves

**Comment by Hanmac** [comment]
@tool4ever can you take a peak on my refactoring to see if I'm on the right track?

Especially the LoseLife/DealDamage one
currently, AI is just taking the lifeloss without reaction

**Comment by Hanmac** [comment]
or @Agetian , what do you think about my WIP?

**Comment by tool4ever** [comment]
Hmn, while I think this was rather low priority logic seems ok?

**Comment by Hanmac** [comment]
> Hmn, while I think this was rather low priority logic seems ok?

That is one of the parts I'm tying to port into Ai Effects,
another part would the logic for Additional Costs, so AI has more control over when to Pay Costs like Kicker, and "Gift a Fish"

**Comment by Agetian** [comment]
Yeah, definitely a good effort, and the implementation looks good to me so far! :+1:

**Comment by Hanmac** [comment]
@Agetian especially these two:
* may draw. If you do, discard
* may discard. If you do, draw

might need more love

**Comment by Hanmac** [comment]
@tool4ever can you use this MR to add the PayEnergy Logic Problem we had this week?

i try to move the last piece of "WillAttack" logic, but i didn't had time yet

**Comment by tool4ever** [comment]
I don't really know how to fix that currently either :/

File: forge-ai/src/main/java/forge/ai/PlayerControllerAi.java [hunk 1]
```diff
@@ -1222,19 +1222,19 @@ public String chooseProtectionType(String string, SpellAbility sa, List<String>
...
+            emptyAbility.setTriggeringObjects(sa.getTriggeringObjects());
+            emptyAbility.setReplacingObjects(sa.getReplacingObjects());
+            emptyAbility.setTrigger(sa.getTrigger());
+            emptyAbility.setReplacementEffect(sa.getReplacementEffect());
+            emptyAbility.setSVars(sa.getSVars());
+            emptyAbility.setCardState(sa.getCardState());
+            emptyAbility.setXManaCostPaid(sa.getRootAbility().getXManaCostPaid());
+            emptyAbility.setTargets(sa.getTargets().clone());
+
             boolean result = ComputerUtil.playNoStack(player, emptyAbility, getGame(), true); // AI needs something to resolve to pay that cost
```
Comment (Hanmac) [inline forge-ai/src/main/java/forge/ai/PlayerControllerAi.java:1238]: @tool4ever can we find a way for AI to pay the Cost without the Need of an extra empty AbilityStatic ?

Using EmptySa also wouldn't fix the problem.

**Comment by Hanmac** [comment]
@tool4ever @Agetian i changed how AI pays for UnlessCost, no need for extra emptyAbility.

probably need some intense testing?

File: forge-gui/src/main/java/forge/player/HumanPlaySpellAbility.java [hunk 2]
```diff
@@ -109,7 +109,9 @@ public final boolean playAbility(final boolean mayChooseTargets, final boolean i
             ability.setPaidLife(0);
         }
 
-        ability = GameActionUtil.addExtraKeywordCost(ability);
+        if (ability.isSpell() && !c.isCopiedSpell()) {
```
Comment (tool4ever) [inline forge-gui/src/main/java/forge/player/HumanPlaySpellAbility.java:112]: this is redundant as addExtraKeywordCost already checks it at the start

Comment (Hanmac) [reply forge-gui/src/main/java/forge/player/HumanPlaySpellAbility.java:112]: Yeah, just to be sure in case I might touch that in the future

**Comment by tool4ever** [comment]
We could change Echo + Cumulative upkeep in SacrificeEffect to use `payCostToPreventEffect`?
technically they are UnlessCost so later logic when not to pay can be added 🤔

**Comment by Hanmac** [comment]
> We could change Echo + Cumulative upkeep in SacrificeEffect to use `payCostToPreventEffect`? technically they are UnlessCost so later logic when not to pay can be added 🤔

that is a bit more tricky, because Echo has an extra trigger, but you could try it.
And like with: 
> You may pay {0} rather than pay the echo cost for permanents you control.

And for Cumulative upkeep, there is also this: "Spend this mana only to pay cumulative upkeep costs."

File: forge-ai/src/main/java/forge/ai/ability/DrawAi.java [hunk 3]
```diff
@@ -547,4 +549,53 @@ public boolean confirmAction(Player player, SpellAbility sa, PlayerActionConfirm
...
+            return false;
+        }
+
+        // Risk Factor Effects
+        for (Player p : AbilityUtils.getDefinedPlayers(host, sa.getParam("Defined"), sa)) {
+            if (p.isOpponentOf(payer)) {
+                if (!p.canDraw()) {
+                    return false;
+                }
+                if (cost.hasSpecificCostType(CostDamage.class)) {
```
Comment (tool4ever) [inline forge-ai/src/main/java/forge/ai/ability/DrawAi.java:575]: all these extra parts checks don't make sense to me

they should already be covered by the `super` anyway

Comment (Hanmac) [reply forge-ai/src/main/java/forge/ai/ability/DrawAi.java:575]: The generic super doesn't check for Damage in comparison to the amount of cards being drawn

File: forge-ai/src/main/java/forge/ai/PlayerControllerAi.java [hunk 4]
```diff
@@ -1220,26 +1221,13 @@ public String chooseProtectionType(String string, SpellAbility sa, List<String>
...
-        emptyAbility.setXManaCostPaid(sa.getRootAbility().getXManaCostPaid());
-        emptyAbility.setTargets(sa.getTargets().clone());
-
-        if (ComputerUtilCost.willPayUnlessCost(sa, player, cost, alreadyPaid, allPayers)) {
-            boolean result = ComputerUtil.playNoStack(player, emptyAbility, getGame(), true); // AI needs something to resolve to pay that cost
-            if (!emptyAbility.getPaidHash().isEmpty()) {
-                // report info to original sa (Argentum Masticore)
-                sa.setPaidHash(emptyAbility.getPaidHash());
+        if (SpellApiToAi.Converter.get(sa.getApi()).willPayUnlessCost(sa, player, cost, alreadyPaid, allPayers)) {
+            if (!ComputerUtilCost.canPayCost(cost, sa, player, true)) {
```
Comment (tool4ever) [inline forge-ai/src/main/java/forge/ai/PlayerControllerAi.java:1225]: you missed this place:
image

AI now cheats on the mana

Comment (Hanmac) [reply forge-ai/src/main/java/forge/ai/PlayerControllerAi.java:1225]: like this?

**Comment by Hanmac** [comment]
@tool4ever can we look about this?

```
        if (cost == null) {
            // Is this fork even used for anything anymore?
            if (ComputerUtilMana.payManaCost(ai, sa, false)) {
                game.getStack().addAndUnfreeze(sa);
                return true;
            }
        } else {
```

that doesn't feel right imo, can this be removed?

File: forge-ai/src/main/java/forge/ai/PlayerControllerAi.java (hunk 4 above)
Comment (tool4ever) [reply forge-ai/src/main/java/forge/ai/PlayerControllerAi.java:1225]: extra object shouldn't be needed but seems to work for now

**Comment by tool4ever** [comment]
I can't think of any case for it.
The Daxos TODO also seems outdated.

**Review by tool4ever — APPROVED** [review]
good enough for now?

**Pull request merged**